src/adapter/src/coord/sequencer/inner.rs

```diff
@@ -2034,9 +2034,68 @@ impl Coordinator {
         })
     }
 
-    async fn peek_stage_optimize(
-        &mut self,
+    async fn peek_stage_optimize(&mut self, ctx: ExecuteContext, stage: PeekStageOptimize) {
+        // Generate data structures that can be moved to another task where we will perform possibly
+        // expensive optimizations.
+        let catalog = self.owned_catalog();
+        let compute = ComputeInstanceSnapshot::new(&self.controller, stage.cluster_id)
+            .expect("compute instance does not exist");
+        let internal_cmd_tx = self.internal_cmd_tx.clone();
+
+        // TODO: Is there a way to avoid making two dataflow_builders (the second is in
+        // optimize_peek)?
+        let id_bundle = self
+            .dataflow_builder(stage.cluster_id)
+            .sufficient_collections(&stage.source_ids);
```

Comment:
We never update the `validity` struct with these IDs. Do we need to worry about these becoming invalid while optimize is off thread?

Reply:
Nice catch! Extended the dependent ids from this which should catch it. I'm not sure how `sources.depends_on()` and `sufficient_collections()` differ, but this seems safer and is cheap.

Reply:
I believe that `sufficient_collections()` includes indexes while `source.depends_on()` does not.